text/0000-rdylib.md

```diff
@@ -0,0 +1,152 @@
+`dylib` output format from what it is today is a breaking change. For example
+metadata will not be present and symbols will be hidden. As a result, this RFC
+has a...
+
+### Transition Plan
+
+This RFC is technically a breaking change, but it is expected to not actually
+break many work flows in practice because there is only one known user of
+rdylibs, the compiler itself. This notably means that plugins will also need to
+be compiled differently, but because they are nightly-only we've got some more
+leeway around them.
+
+All other known users of the `dylib` output crate type fall into the cdylib use
+case. The "breakage" here would mean:
+
+* The metadata section no longer exists. In almost all cases this just means
+  that the output artifacts will get smaller if it isn't present, it's expected
+  that no one other than the compiler itself is actually consuming this
+  information.
+* Rust symbols will be hidden by default. The symbols, however, have
+  unpredictable hashes so there's not really any way they can be meaningfully
+  leveraged today.
+
+Given that background, it's expected that if there's a smooth migration path for
+plugins and the compiler then the "breakage" here won't actually appear in
+practice. The proposed implementation strategy and migration path is:
+
+1. Implement the `rdylib` output type as proposed in this RFC.
+2. Change Cargo to use `--crate-type rdylib` when compiling plugins instead of
+   `--crate-type dylib` + `-C prefer-dynamic`.
+3. Implement the changes to the `dylib` output format as proposed in this RFC.
+
+So long as the steps are spaced apart by a few days it should be the case that
+no nightly builds break if they're always using an up-to-date nightly compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust's ephemeral and ill-defined "linkage model" is... well... ill defined and
+ephemeral. This RFC is an extension of this model, but it's difficult to reason
+about extending that which is not well defined. As a result there could be
+unforseen interactions between this output format and where it's used.
+
+As usual, of course, proposing a breaking change is indeed a drawback. It is
+expected that RFC doesn't break anything in practice, but that'd be difficult to
+gauge until it's implemented.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* Instead of reinterpreting the `dylib` output format as a cdylib, we could
+  continue interpreting it as an rdylib and add a new dedicated `cdylib` output
+  format. This would not be a breaking change, but it doesn't come without its
+  drawbacks. As the most common output type, many projects would have to switch
+  to `cdylib` from `dylib`, meaning that they no longer support older Rust
+  compilers. This may also take time to propagate throughout the community. It's
+  also arguably a "better name", so this RFC proposes an
+  in-practice-not-a-breaking-change by adding a worse name of `rdylib` for the
+  less used output format.
```

Comment:
In the distant future with a stable Rust ABI, this format might eventually become dominant. That said, `rdylib` reads as "Rust dynamic library", so no real harm done here.